### PR TITLE
Add `remove_if_expression` rule

### PR DESCRIPTION
**Pull request by jiwonz**
Add `remove_if_expression` rule which removes luau's if expression syntax from source then translate it into closure and if statement.

Tests are succeeded.

# Examples
## Input
```luau
local a = if true then 1 else 2
```
## Output
```lua
local a = (true and {(1)} or {(2)})[1]
```

### CODE REVIEW

File: src/rules/remove_if_expression.rs [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+#[derive(Default)]
+struct Processor {}
+
+impl NodeProcessor for Processor {
+    fn process_expression(&mut self, expression: &mut Expression) {
+        let call_exp: Option<Expression> = if let Expression::If(if_exp) = expression {
+            let result_return = nodes::ReturnStatement::one(if_exp.get_result().clone());
+            let else_result_return = nodes::ReturnStatement::one(if_exp.get_else_result().clone());
+            let front_block = nodes::Block::new(vec![], Some(result_return.into()));
+            let front_branch = nodes::IfBranch::new(if_exp.get_condition().clone(), front_block);
```
Comment (jeparlefrancais) [inline src/rules/remove_if_expression.rs:18]: You can take advantage that Statement and LastStatement can be made into a Block, most methods will accept `Into<Block>` (or `Into<Statement>` or `Into<Expression>`).

```rust
let front_branch = IfBranch::new(if_exp.get_condition().clone(), result_return);
```

File: tests/rule_tests/remove_if_expression.rs [hunk 2]
```diff
@@ -0,0 +1,23 @@
+use darklua_core::rules::{RemoveIfExpression, Rule};
+
+test_rule!(
+    remove_if_expression,
+    RemoveIfExpression::default(),
+    assign_if_expression("local a = if true then 1 else 2") => "local a = (function() if true then return 1 else return 2 end end)()",
+	assign_if_expression_with_elseif("local a = if true then 1 elseif false then 2 else 3") => "local a = (function() if true then return 1 elseif false then return 2 else return 3 end end)()"
+);
```
Comment (jeparlefrancais) [inline tests/rule_tests/remove_if_expression.rs:8]: Let's add more tests! For example, we need to treat function calls and `...` in a specific way to make sure we are not returning multiple values. For that we need to wrap them in parentheses.

**Review by jeparlefrancais — COMMENTED** [review]
Thank for contributing! I think we could switch the replacement strategy though. Someone in the Roblox open source discord shared this trick where you can do:

```
local a = if condition() then 1 else 0
local a = (condition() and { 1 } or { 0 })[1]
```

We should also wrap potential expressions that can return multiple values to avoid creating a bigger table than needed. So `...` and functions calls should be wrapped in parentheses!

**Comment by jiwonz** [comment]
> Thank for contributing! I think we could switch the replacement strategy though. Someone in the Roblox open source discord shared this trick where you can do:
> 
> ```
> local a = if condition() then 1 else 0
> local a = (condition() and { 1 } or { 0 })[1]
> ```
> 
> We should also wrap potential expressions that can return multiple values to avoid creating a bigger table than needed. So `...` and functions calls should be wrapped in parentheses!

Hello! Thanks for reviewing! I read your all 3 comments which were really helpful for me to improve my contribution.

but sorry. one thing that I'm not quite understanding is "We should also wrap potential expressions that can return multiple values to avoid creating a bigger table than needed. So `...` and functions calls should be wrapped in parentheses!" I'm not sure what `...` and function calls are actually meaning.

Can you please provide more example codes so can you help me I understand? Sorry about that.

**Comment by jiwonz** [comment]
> > but sorry. one thing that I'm not quite understanding is "We should also wrap potential expressions that can return multiple values to avoid creating a bigger table than needed. So `...` and functions calls should be wrapped in parentheses!" I'm not sure what `...` and function calls are actually meaning.
> > Can you please provide more example codes so can you help me I understand? Sorry about that.
> 
> Of course 😄 Take this example:
> 
> ```lua
> local function oof(...: string)
>     return if condition(...) then ... else transform(...)
> end
> -- converts to
> local function oof(...: string)
>     return (condition() and { (...) } or { transform(...) })[1]
> end
> ```
> 
> 1. Imagine that `oof` is called with multiple arguments, `...` would expand in the table and put more values than needed.
> 2. Imagine that `transform(...)` returns multiple values, it would also expand in the table and put more values than needed again
> 
> The wrapping in parentheses will make the extra values go away.

Thank you so much! It seems possible to implement easily except for `elseif` hell... hm

**Comment by jiwonz** [comment]
This PR may close #207

**Comment by Stefanuk12** [comment]
In my opinion (as stated in #212), this is how the transformation should happen:
```lua
local maxValue = if a > b then a else b
```
goes to
```lua
local maxValue = b
if a > b then
    maxValue = a
end
```

I took a look at the Luau bytecode and that's what it would theoretically decompile to.

**Comment by jiwonz** [comment]
> In my opinion (as stated in #212), this is how the transformation should happen:
> 
> ```lua
> local maxValue = if a > b then a else b
> ```
> 
> goes to
> 
> ```lua
> local maxValue
> if a > b then
>     maxValue = a
> else
>     maxValue = b
> end
> ```
> 
> ## Why
> Using an `if` saves a few instructions and is truer to what it actually Luau compiles the statement to.
> 
> ```
> function anon_0(??)
>     1     GETIMPORT   R2 1      ; a
>     2     GETIMPORT   R3 3      ; b
>     3     JUMPIFNOTLT R3 R2 L0
>     4     NEWTABLE    R1 0 1
>     5     GETIMPORT   R2 1      ; a
>     6     SETLIST     R1 R2 1   ; 1
>     7     JUMPIF      R1 L1
>     8  L0 NEWTABLE    R1 0 1
>     9     GETIMPORT   R2 3      ; b
>     10    SETLIST     R1 R2 1   ; 1
>     11 L1 GETTABLEN   R0 R1 1
>     12    GETIMPORT   R1 5      ; print
>     13    MOVE        R2 R0
>     14    CALL        R1 1 0
>     15    RETURN      R0 0
> end
> 
> function anon_1(??)
>     1     LOADNIL     R0
>     2     GETIMPORT   R1 1      ; a
>     3     GETIMPORT   R2 3      ; b
>     4     JUMPIFNOTLT R2 R1 L0
>     5     GETIMPORT   R0 1      ; a
>     6     JUMP        L1
>     7  L0 GETIMPORT   R0 3      ; b
>     8  L1 GETIMPORT   R1 5      ; print
>     9     MOVE        R2 R0
>     10    CALL        R1 1 0
>     11    RETURN      R0 0
> end
> 
> function anon_2(??)
>     1     GETIMPORT   R1 1      ; a
>     2     GETIMPORT   R2 3      ; b
>     3     JUMPIFNOTLT R2 R1 L0
>     4     GETIMPORT   R0 1      ; a
>     5     JUMP        L1
>     6  L0 GETIMPORT   R0 3      ; b
>     7  L1 GETIMPORT   R1 5      ; print
>     8     MOVE        R2 R0
>     9     CALL        R1 1 0
>     10    RETURN      R0 0
> end
> ```
> 
> ### Samples (in order)
> ```lua
> local maxValue = ((a > b) and {(a)} or {(b)})[1]
> print(maxValue)
> ```
> 
> ```lua
> local maxValue
> if a > b then
>     maxValue = a
> else
>     maxValue = b
> end
> print(maxValue)
> ```
> 
> ```lua
> local maxValue = if a > b then a else b
> print(maxValue)
> ```

It is not that easy to convert if expression into if statement since the case like this code above would be a nightmare of side effects
```lua
local c = ...
local a, c = sideEffect() + if condition then compute(c) else default(), resetC()
```

**Comment by Stefanuk12** [comment]
I'm on mobile but I believe that
```lua
local c = ...
local a, c = sideEffect() + if condition then compute(c) else default(), resetC()
```
compiles back to
```lua
local c = ...
local a = sideEffect() + if condition then compute(c) else default()
local c = resetC()
```
where the results of `sideEffect` and the if expr are stored in separate registers before added up so it's basically equal to
```lua
local c = ...
local a = sideEffect()
a += if condition then compute(c) else default()
local c = resetC()
```

Looking at the bytecode, it's executed as expected in order from left to right. `sideEffect` is callled then stored and finally added to the result of the if expr. Then, `resetC` is called.

Similar to how darklua handles `a[foo()] += 1`, a temporary variable could be made that computes the final value.

However, I can acknowledge how it may be difficult to know when or how to split the expression up. For example, setting `a` equal to the result then setting `a` again but with the if expr result.

I'm not too experienced with these things but Rust should make it easier since you can match Expr and make sure all cases are covered.

**Pull request closed without merging**